File: .agents/skills/evo-asset/scripts/compute_asset_id.js
#!/usr/bin/env node
/**
 * compute_asset_id.js
 *
 * Computes the content-addressable asset_id for an EvoMap evolution asset.
 *
 * Algorithm:
 *   sha256(canonical_json(asset_without_asset_id_field))
 *
 * Canonical JSON = all keys sorted alphabetically at every nesting level,
 * serialized as a compact string (no extra whitespace).
 *
 * Usage:
 *   node compute_asset_id.js <asset.json>
 *
 * Output:
 *   sha256:<hex_digest>
 *
 * Example:
 *   node compute_asset_id.js examples/gene_example.json
 *   sha256:3a7f2c9d...
 */

'use strict';

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// ── Helpers ────────────────────────────────────────────────────────────────

/**
 * Recursively sort all object keys alphabetically (canonical JSON).
 * Arrays preserve their order; only object keys are sorted.
 */
function sortKeysDeep(value) {
  if (Array.isArray(value)) {
    return value.map(sortKeysDeep);
  }
  if (value !== null && typeof value === 'object') {
    return Object.keys(value)
      .sort()
      .reduce((acc, key) => {
        acc[key] = sortKeysDeep(value[key]);
        return acc;
      }, {});
  }
  return value;
}

/**
 * Compute sha256 of a UTF-8 string, return hex digest.
 */
function sha256(str) {
  return crypto.createHash('sha256').update(str, 'utf8').digest('hex');
}

// ── Main ───────────────────────────────────────────────────────────────────

const filePath = process.argv[2];

if (!filePath) {
  console.error('Usage: node compute_asset_id.js <asset.json>');
  console.error('');
  console.error('  Reads the JSON file, removes the "asset_id" field,');
  console.error('  and prints the canonical SHA256 asset_id.');
  process.exit(1);
}

const resolvedPath = path.resolve(filePath);

if (!fs.existsSync(resolvedPath)) {
  console.error(`Error: File not found: ${resolvedPath}`);
  process.exit(1);
}

let raw;
try {
  const content = fs.readFileSync(resolvedPath, 'utf8');
  raw = JSON.parse(content);
} catch (err) {
  console.error(`Error: Failed to parse JSON — ${err.message}`);
  process.exit(1);
}

if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
  console.error('Error: Input must be a JSON object (not an array or primitive).');
  process.exit(1);
}

// Remove asset_id field before hashing
const { asset_id: _removed, ...assetWithoutId } = raw;

// Canonical JSON: sort keys recursively, compact serialization
const canonical = JSON.stringify(sortKeysDeep(assetWithoutId));

const digest = sha256(canonical);
console.log(`sha256:${digest}`);

File: .agents/skills/evo-asset/scripts/compute_asset_id.py
#!/usr/bin/env python3
"""
compute_asset_id.py

Computes the content-addressable asset_id for an EvoMap evolution asset.

Algorithm:
    sha256(canonical_json(asset_without_asset_id_field))

Canonical JSON = all keys sorted alphabetically at every nesting level,
serialized as a compact string (no extra whitespace).

Usage:
    python compute_asset_id.py <asset.json>

Output:
    sha256:<hex_digest>

Example:
    python compute_asset_id.py examples/gene_example.json
    sha256:3a7f2c9d...

Requirements:
    Python 3.6+ (standard library only, no pip install needed)
"""

import sys
import json
import hashlib


def sort_keys_deep(value):
    """
    Recursively sort all dict keys alphabetically (canonical JSON).
    Lists preserve their order; only dict keys are sorted.
    """
    if isinstance(value, list):
        return [sort_keys_deep(item) for item in value]
    if isinstance(value, dict):
        return {k: sort_keys_deep(v) for k, v in sorted(value.items())}
    return value


def compute_asset_id(asset_dict: dict) -> str:
    """
    Compute the canonical SHA256 asset_id for an asset.

    Steps:
      1. Remove 'asset_id' field (if present)
      2. Sort all keys recursively
      3. Serialize to compact JSON string
      4. UTF-8 encode and SHA256 hash
      5. Return 'sha256:' + hex_digest
    """
    # Remove asset_id field
    without_id = {k: v for k, v in asset_dict.items() if k != 'asset_id'}

    # Canonical JSON: sorted keys, compact (no spaces)
    canonical = json.dumps(
        sort_keys_deep(without_id),
        separators=(',', ':'),
        ensure_ascii=False
    )

    # SHA256
    digest = hashlib.sha256(canonical.encode('utf-8')).hexdigest()
    return f'sha256:{digest}'


def main():
    if len(sys.argv) < 2:
        print('Usage: python compute_asset_id.py <asset.json>', file=sys.stderr)
        print('', file=sys.stderr)
        print('  Reads the JSON file, removes the "asset_id" field,', file=sys.stderr)
        print('  and prints the canonical SHA256 asset_id.', file=sys.stderr)
        sys.exit(1)

    file_path = sys.argv[1]

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            raw = json.load(f)
    except FileNotFoundError:
        print(f'Error: File not found: {file_path}', file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f'Error: Failed to parse JSON — {e}', file=sys.stderr)
        sys.exit(1)

    if not isinstance(raw, dict):
        print('Error: Input must be a JSON object (not an array or primitive).', file=sys.stderr)
        sys.exit(1)

    result = compute_asset_id(raw)
    print(result)


if __name__ == '__main__':
    main()
